### PR TITLE
Add ToDecibels CPU kernel

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

#### Why we need this PR?
- Some magnitudes are often represented in dB. We need a kernel for that

#### What happened in this PR?
- Implemented ToDecibels kernel for the CPU
- Added Kernel unit tests

**JIRA TASK**: [DALI-1173]

### CODE REVIEW

File: dali/kernels/signal/decibel/decibel_calculator.h [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+namespace dali {
+namespace kernels {
+namespace signal {
+
+template <typename T>
+struct DecibelCalculator {
+ public:
+  explicit DecibelCalculator(T mul = 10.0, T s_ref = 1.0, T min_ratio = 1e-8)
+      : mul_log2_(mul / std::log2(10.0))
```
Comment (mzient) [inline dali/kernels/signal/decibel/decibel_calculator.h]:
```suggestion
      : mul_log2_(mul * std::log10(2.0))
```
...and we could have this as a constant - it's a pity it's not defined in math.h

File: dali/kernels/signal/decibel/decibel_calculator.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+      , inv_s_ref_(s_ref == 1.0 ? 1.0 : 1.0 / s_ref)
+      , min_ratio_(min_ratio) {
+    assert(min_ratio_ > 0.0);
+    assert(inv_s_ref_ > 0.0);
+  }
+
+  T operator()(T s) const {
```
Comment (mzient) [inline dali/kernels/signal/decibel/decibel_calculator.h]: inline? Maybe even DALI_FORCEINLINE?

File: dali/kernels/signal/decibel/to_decibels_cpu.cc [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    const InTensorCPU<T, Dims> &in,
+    const ToDecibelsArgs<T> &args) {
+  auto out_shape = in.shape;
+  auto data_size = volume(in.shape);
+  std::vector<TensorShape<DynamicDimensions>> tmp = {out_shape};  // workaround for clang-6 bug
+  KernelRequirements req;
+  req.output_shapes = {TensorListShape<DynamicDimensions>(tmp)};
+
+  s_ref_ = args.s_ref;
+  if (args.ref_max) {
```
Comment (JanuszL) [inline dali/kernels/signal/decibel/to_decibels_cpu.cc]: I'm not very keen we must iterate over the whole data in setup. I would consider setup to be something light. Can we move it to run?

Comment (jantonguirao) [reply dali/kernels/signal/decibel/to_decibels_cpu.cc]: Sure. I actually had it in Run before :)

File: dali/kernels/signal/decibel/decibel_calculator.h (hunk 2 above)
Comment (jantonguirao) [reply dali/kernels/signal/decibel/decibel_calculator.h]: done

File: dali/kernels/signal/decibel/decibel_calculator.h (hunk 1 above)
Comment (jantonguirao) [reply dali/kernels/signal/decibel/decibel_calculator.h]: done

File: dali/kernels/signal/decibel/to_decibels_cpu.cc (hunk 3 above)
Comment (jantonguirao) [reply dali/kernels/signal/decibel/to_decibels_cpu.cc]: Done

File: dali/kernels/signal/decibel/to_decibels_cpu.cc [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  auto in_size = volume(in.shape);
+  auto out_size = volume(out.shape);
+  assert(out_size == in_size);
+
+  s_ref_ = args.s_ref;
+  if (args.ref_max) {
+    s_ref_ = 0.0;
+    for (int64_t i = 0; i < in_size; i++) {
+      if (in.data[i] > s_ref_)
+        s_ref_ = in.data[i];
```
Comment (JanuszL) [inline dali/kernels/signal/decibel/to_decibels_cpu.cc]: This s_ref_  probably no longer needs to be a member.

Comment (jantonguirao) [reply dali/kernels/signal/decibel/to_decibels_cpu.cc]: Right, will fix

File: dali/kernels/signal/decibel/to_decibels_cpu.cc [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+namespace kernels {
+namespace signal {
+
+template <typename T, int Dims>
+ToDecibelsCpu<T, Dims>::~ToDecibelsCpu() = default;
+
+template <typename T, int Dims>
+KernelRequirements ToDecibelsCpu<T, Dims>::Setup(
+    KernelContext &context,
+    const InTensorCPU<T, Dims> &in,
```
Comment (szalpal) [inline dali/kernels/signal/decibel/to_decibels_cpu.cc]: shouldn't it be `const InTensorCPT<const T, Dims>`?

File: dali/kernels/signal/decibel/to_decibels_cpu.h [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+namespace dali {
+namespace kernels {
+namespace signal {
+
+template <typename T = float, int Dims = 1>
+class DLL_PUBLIC ToDecibelsCpu {
+ public:
+  static_assert(std::is_floating_point<T>::value,
+    "Only floating point types are supported");
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/kernels/signal/decibel/to_decibels_cpu.h]: I'm only defining the kernel for `float` and `double`. It's what makes sense for now

File: dali/kernels/signal/decibel/to_decibels_cpu.cc [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+template class ToDecibelsCpu<float, 1>;
+template class ToDecibelsCpu<float, 2>;
+template class ToDecibelsCpu<float, 3>;
+template class ToDecibelsCpu<float, 4>;
+
+template class ToDecibelsCpu<double, 1>;
+template class ToDecibelsCpu<double, 2>;
+template class ToDecibelsCpu<double, 3>;
+template class ToDecibelsCpu<double, 4>;
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/kernels/signal/decibel/to_decibels_cpu.cc]: 1. Keep implementation in the *.cc
2. Avoid compiling stuff twice
3. Keep control of what we instantiate

File: dali/kernels/signal/decibel/decibel_calculator.h [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+#include "dali/core/force_inline.h"
+
+namespace dali {
+namespace kernels {
+namespace signal {
+
+template <typename T>
+struct DecibelCalculator {
+ public:
+  explicit DecibelCalculator(T mul = 10.0, T s_ref = 1.0, T min_ratio = 1e-8)
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/kernels/signal/decibel/decibel_calculator.h]: It's not. You can invoke this constructor with one argument only (linter complained)

File: dali/kernels/signal/decibel/to_decibels_cpu.cc (hunk 5 above)
Comment (jantonguirao) [reply dali/kernels/signal/decibel/to_decibels_cpu.cc]: No
```
template <typename StorageBackend, typename ElementType, int sample_dim>
using InTensor = TensorView<StorageBackend, const ElementType, sample_dim>;

template <typename ElementType, int sample_dim>
using InTensorCPU = InTensor<StorageCPU, ElementType, sample_dim>;
```

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build

**Pull request merged**